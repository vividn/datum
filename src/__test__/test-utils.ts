import { CouchDbError } from "../errors";
import {
  DataOnlyDocument,
  DatumData,
  DatumDocument,
  DatumMetadata,
  EitherDocument,
  EitherPayload,
} from "../documentControl/DatumDocument";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "../time/parseTimeStr";
import { now } from "../time/timeUtils";
import { connectDb } from "../auth/connectDb";
import * as connectDbModule from "../auth/connectDb";
import { defaultIdComponents } from "../ids/defaultIdComponents";
import { buildIdStructure } from "../ids/buildIdStructure";
import { defaults } from "../input/defaults";
import { assembleId } from "../ids/assembleId";
import * as newHumanIdModule from "../meta/newHumanId";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const mockDocMissingError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing",
  message: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  status: 404,
  message: "deleted",
  name: "not_found",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing_named_view",
  message: "missing_named_view",
};

export async function resetTestDb(
  db: PouchDB.Database & {
    __opts?: PouchDB.Configuration.DatabaseConfiguration;
    _destroyed?: boolean;
    _closed?: boolean;
  },
): Promise<PouchDB.Database> {
  await db.destroy().catch(pass);
  // nasty hack to reopen closed database
  delete db._destroyed;
  delete db._closed;
  return db.constructor(db.name, db.__opts);
}

export function testDbLifecycle(
  dbName: string,
): PouchDB.Database<EitherPayload> {
  const db = connectDb({ db: dbName });

  beforeEach(async () => {
    await resetTestDb(db);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db);
  });

  afterEach(async () => {
    await db.destroy().catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr, referenceTime: DateTime.now() });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => Promise<O>,
): (...args: A) => Promise<O> {
  return async (...args: A): Promise<O> => {
    pushNow(timeStr);
    const returnVal = await fn(...args);
    popNow();
    return returnVal;
  };
}

// TODO: Transition all tests to use makeDoc where appropriate
export function makeDoc(
  data: DatumData,
  meta: DatumMetadata | false = {},
  include_rev = false,
): EitherDocument {
  let doc: EitherDocument;
  if (meta === false) {
    doc = { ...data } as DataOnlyDocument;
  } else {
    doc = { ...data, meta } as DatumDocument;
  }

  meta = meta || {};
  if (doc._id === undefined) {
    let idStructure: string;
    if (meta?.idStructure) {
      idStructure = meta.idStructure;
    } else {
      const { defaultIdParts, defaultPartitionParts } = defaultIdComponents({
        data,
      });
      idStructure = buildIdStructure({
        idParts: defaultIdParts,
        delimiter: defaults.idDelimiter,
        partition: defaultPartitionParts,
      });
    }
    doc._id = assembleId({ payload: doc, idStructure });
  }

  if (include_rev) {
    doc._rev = "1-foo";
  }
  return doc;
}

export async function deterministicHumanIds(seed?: number): Promise<void> {
  let a = seed || 20231018;
  function random() {
    const x = Math.sin(a++) * 10000;
    return x - Math.floor(x);
  }
  function mockNewHumanId(): string {
    return random().toString(36).slice(2) + random().toString(36).slice(2);
  }

  beforeEach(() => {
    a = seed || 20231018;
    jest
      .spyOn(newHumanIdModule, "newHumanId")
      .mockImplementation(mockNewHumanId);
  });
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
