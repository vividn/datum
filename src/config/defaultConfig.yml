# Datum Configuration

# the directory where you will store you datum views, specs, and other (version controlled) files to setup your database
project_dir: "%HOME%/datum"

# default database to use 
# override with --db
db: "datum"

connection:
  # host can be 
  #   - a URL of a couchdb instance, e.g. http://localhost:5984
  #   - a root directory where to store db files with pouchdb, use %DATA% for XDG_DATA_HOME
  # Override with $COUCHDB_HOST or with --host
  host: "%DATA%/datum"

  # default user to use for couchdb instances. Has no effect on pouchdb
  # Override with $COUCHDB_USER or with --user
  user: "user"

  # password to use for couchdb instances. Has no effect on pouchdb
  # Override wtih $COUCHDB_PASSWORD or with --password
  password: null
