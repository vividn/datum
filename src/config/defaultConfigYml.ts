import yaml from "yaml";

export const defaultConfigDir =
  process.env["XDG_CONFIG_HOME"] || `${process.env["HOME"]}/.config`;
export const defaultConfigPath = `${defaultConfigDir}/datum/datumrc.yml`;

export const defaultConfigYml = yaml.parseDocument(`# Datum Configuration

# default database to use 
# override with --db
db: "datum"

# host can be 
#   - a root directory where to store db files with pouchdb, use %DATA% for XDG_DATA_HOME
#   - a URL of a couchdb instance, e.g. http://localhost:5984
# Override with $DATUM_HOST or with --host
host: "%DATA%/datum"

# For couchdb connnections you can specify a user and password here, override with $COUCHDB_USER, $COUCHDB_PASSWORD or --user, --password
# Has no effect for pouchdb connections
# Leave blank or null to prompt
# user: null
# password: null
`);
