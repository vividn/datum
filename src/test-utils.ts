import Nano, { DocumentScope } from "nano";
import { CouchDbError } from "./errors";
import { EitherPayload } from "./documentControl/DatumDocument";
import * as connectDb from "./auth/connectDb";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "./time/parseTimeStr";
import { now } from "./time/timeUtils";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const testNano = Nano(`http://admin:password@localhost:5983`);

export const mockDocMissingError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing",
  error: "not_found",
  reason: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "deleted",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing_named_view",
  error: "not_found",
  reason: "missing_named_view",
};

export async function resetTestDb(
  dbName: string
): Promise<DocumentScope<EitherPayload>> {
  const maxTries = 3;
  let tries = 0;
  await testNano.db.destroy(dbName).catch(pass);
  while (tries++ <= maxTries) {
    await testNano.db.destroy(dbName).catch(pass);
    await testNano.db.create(dbName).catch(pass);
    if ((await testNano.db.list()).includes(dbName)) {
      return testNano.use(dbName) as DocumentScope<EitherPayload>;
    }
  }
  throw Error(`Unable to reset database after ${maxTries} attempts`);
}

export function testDbLifecycle(dbName: string): DocumentScope<EitherPayload> {
  const db = testNano.use(dbName) as DocumentScope<EitherPayload>;

  beforeEach(async () => {
    await resetTestDb(dbName);
    jest.spyOn(connectDb, "connectDb").mockImplementation(async () => db);
  });

  afterEach(async () => {
    await testNano.db.destroy(dbName).catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => O
): (...args: A) => O {
  return (...args: A): O => {
    pushNow(timeStr);
    const returnVal: O = fn(...args);
    popNow();
    return returnVal;
  };
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
