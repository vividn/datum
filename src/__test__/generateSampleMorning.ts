import { endCmd } from "../commands/endCmd";
import { occurCmd } from "../commands/occurCmd";
import { startCmd } from "../commands/startCmd";
import { switchCmd } from "../commands/switchCmd";
import { popNow, pushNow, setNow } from "./test-utils";

export async function generateSampleMorning(date: string): Promise<void> {
  pushNow(`8:30 ${date}`);
  await endCmd("sleep");
  setNow("+5");
  await startCmd("sleep -y -t 2320");
  setNow("8:40");
  await switchCmd("project german");
  await switchCmd("text fiction_book");
  setNow("9:10");
  await endCmd("project");
  await endCmd("text");
  setNow("9:30");
  await switchCmd("environment outside");
  await startCmd("stretch");
  setNow("+7");
  await endCmd("stretch");
  await startCmd("run");
  setNow("+30");
  await endCmd("run distance=5.4");
  await startCmd("stretch");
  setNow("+8");
  await endCmd("stretch");
  await switchCmd("environment home");
  setNow("+3");
  await occurCmd("pushup amount=10");
  setNow("10:30");
  await switchCmd("environment outside 5");
  setNow("11");
  await occurCmd("caffeine amount=100 -c coffee");
  popNow();
}
