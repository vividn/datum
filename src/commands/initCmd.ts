import { promises as fs } from "fs";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { initConfig } from "../config/initConfig";
import { MainDatumArgs } from "../input/mainArgs";

export type InitCmdArgs = MainDatumArgs & {
  overwrite?: boolean;
  nonInteractive?: boolean;
};

export const initCmdArgs = new ArgumentParser({
  description: "initialize the datum configuration",
  prog: "datum init",
  usage: `%(prog)s [--overwrite] [--global]`,
});

initCmdArgs.add_argument("--overwrite", {
  help: "overwrite the existing configuration file",
  action: "store_true",
});
initCmdArgs.add_argument("--non-interactive", {
  help: "do not prompt for input",
  action: "store_true",
});

export async function initCmd(
  args: InitCmdArgs | string | string[],
  preparsed?: Partial<InitCmdArgs>,
): Promise<void> {
  args = parseIfNeeded(initCmdArgs, args, preparsed);

  const xdgConfig =
    process.env.XDG_CONFIG_HOME ?? `${process.env.HOME}/.config`;
  const configDir = `${xdgConfig}/datum`;

  await fs.mkdir(configDir, { recursive: true });
  const filepath = `${configDir}/datumrc.yml`;
  let fileExists;
  try {
    await fs.access(filepath);
    fileExists = true;
  } catch {
    fileExists = false;
  }

  if (fileExists && !args.overwrite) {
    throw new Error(
      `datumrc.yml already exists at ${filepath}. Use --overwrite to overwrite.`,
    );
  }

  const newConfig = await initConfig(args);
  await fs.writeFile(filepath, newConfig.toString(), { encoding: "utf-8" });
}
