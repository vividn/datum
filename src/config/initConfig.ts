import prompts, { PromptObject } from "prompts";
import yaml from "yaml";
import fs from "fs";
import { InitCmdArgs } from "../commands/initCmd";

export async function initConfig(args: InitCmdArgs): Promise<yaml.Document> {
  const newConfig = yaml.parseDocument(
    fs.readFileSync(__dirname + "/defaultConfig.yml", "utf8"),
  );

  // prompts.override({ projectDir: "abcde" });

  let isLocalCouchRunning;
  try {
    await fetch("http://localhost:5984");
    isLocalCouchRunning = true;
  } catch {
    isLocalCouchRunning = false;
  }

  const defaults = {
    projectDir: args.projectDir ?? (newConfig.get("project_dir") as string),
    dbType: isLocalCouchRunning ? "couchdb" : "pouchdb",
    host:
      args.host ??
      process.env["COUCHDB_HOST"] ??
      (isLocalCouchRunning ? "http://localhost:5984" : "%DATA%/datum"),
    user:
      args.user ??
      process.env["COUCHDB_USER"] ??
      (newConfig.getIn(["connection", "user"]) as string | null),
    password: null,
    db: args.db ?? (newConfig.get("db") as string),
  };

  const questions: PromptObject[] = [
    {
      name: "projectDir",
      type: "text",
      message:
        "Project directory. This is where you will set up and maintain your custom views, specs, hardcoded data documents, and other files. Datum then uses this directory to automatically setup each database. It is recommended to put this under version control.",
      initial: defaults.projectDir,
    },
    {
      name: "dbType",
      type: "select",
      message: "Database type",
      choices: [
        { title: "CouchDB", value: "couchdb" },
        { title: "PouchDB", value: "pouchdb" },
      ],
      initial: isLocalCouchRunning ? 0 : 1,
    },
    {
      name: "host",
      type: "text",
      message: (_, values) =>
        values.dbType === "couchdb"
          ? "CouchDB location with port"
          : "PouchDB database root location. %DATA% is the XDG_DATA_HOME directory",
      initial: (_, values) =>
        args.host ??
        (process.env["COUCHDB_HOST"] ||
          (values.dbType === "couchdb"
            ? "http://localhost:5984"
            : "%DATA%/datum")),
    },
    {
      name: "user",
      type: (_, values) => (values.dbType === "couchdb" ? "text" : null),
      message: "Default CouchDB username",
      initial: defaults.user ?? "user",
    },
    {
      name: "password",
      type: (_, values) => (values.dbType === "couchdb" ? "password" : null),
      message:
        "Couchdb password. WARNING: entering here will store in plain text in the config file. Leave blank to prompt at runtime. Can also use $COUCHDB_PASSWORD or --password",
    },
    {
      name: "db",
      type: "text",
      message: "Default database name",
      initial: "datum",
    },
  ];

  const answers = args.nonInteractive
    ? defaults
    : await prompts(questions, { onCancel: () => process.exit(1) });

  newConfig.set("project_dir", answers.projectDir);
  newConfig.set("db", answers.db);
  newConfig.setIn(["connection", "host"], answers.host);
  if (answers.user) {
    newConfig.setIn(["connection", "user"], answers.user);
  }
  if (answers.password) {
    newConfig.setIn(["connection", "password"], answers.password);
  }

  return newConfig;
}
