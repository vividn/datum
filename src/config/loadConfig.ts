import yaml from "yaml";
import { MainDatumArgs } from "../input/mainArgs";
import { initConfig } from "./initConfig";
import { defaultConfigPath } from "./defaultConfigYml";

export type DatumConfig = {
  db?: string;
  host?: string;
  user?: string;
  password?: string;
};

export function loadConfig(args: MainDatumArgs): DatumConfig {
  const fs = eval("require('fs')");
  const configFile = args.configFile ?? defaultConfigPath;
  let config: DatumConfig;
  try {
    config = yaml
      .parseDocument(fs.readFileSync(configFile, "utf8"))
      .toJSON() as DatumConfig;
  } catch (e: any) {
    if (e.code === "ENOENT") {
      if (args.configFile) {
        throw new Error(`Config file not found: ${args.configFile}`);
      } else {
        console.info(`Welcome to datum!`);
        console.info(`Creating a configuration file at ${configFile}`);
        console.info("");
        config = initConfig();
      }
    } else {
      throw e;
    }
  }

  if (config?.host) {
    config.host = config.host.replaceAll(/%HOME%|~/g, process.env["HOME"]!);
    config.host = config.host.replaceAll(
      "%DATA%",
      process.env["XDG_DATA_HOME"] ?? process.env["HOME"] + "/.local/share",
    );
  }

  return config;
}
