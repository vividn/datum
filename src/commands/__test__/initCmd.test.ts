import fs from "fs";
import { initCmd } from "../initCmd";
import { pass } from "../../__test__/test-utils";
import yaml from "yaml";
import { defaultConfigYml } from "../../config/defaultConfigYml";

describe("initCmd", () => {
  beforeEach(() => {
    process.env.XDG_CONFIG_HOME = "/tmp";
    delete process.env.COUCHDB_HOST;
    delete process.env.COUCHDB_USER;
    delete process.env.COUCHDB_PASSWORD;

    try {
      fs.rmSync("/tmp/datum/datumrc.yml");
    } catch (e) {
      pass;
    }

    // initConfig uses fetch to query if couchdb is running, so mock it
    jest
      .spyOn(global, "fetch")
      .mockImplementation(() => Promise.reject(new Error("url not found")));
  });

  it("writes datumrc into the xdg_config directory", async () => {
    process.env.XDG_CONFIG_HOME = "/tmp";
    await initCmd({ nonInteractive: true });
    fs.accessSync("/tmp/datum/datumrc.yml");
    const configFile = fs.readFileSync("/tmp/datum/datumrc.yml", "utf-8");
    expect(configFile).toEqual(defaultConfigYml);
  });

  it("fails if datumrc exists and --overwrite is not specified", async () => {
    process.env.XDG_CONFIG_HOME = "/tmp";
    fs.writeFileSync("/tmp/datum/datumrc.yml", "test");
    await expect(initCmd({ nonInteractive: true })).rejects.toThrow();
  });

  it("overwrites datumrc if it exists and --overwrite is specified", async () => {
    process.env.XDG_CONFIG_HOME = "/tmp";
    fs.writeFileSync("/tmp/datum/datumrc.yml", "test");
    await initCmd({ nonInteractive: true, overwrite: true });
    fs.accessSync("/tmp/datum/datumrc.yml");
    const configFile = fs.readFileSync("/tmp/datum/datumrc.yml", "utf-8");
    expect(configFile).toEqual(defaultConfigYml);
  });

  it("uses environment variables if they are set", async () => {
    process.env.XDG_CONFIG_HOME = "/tmp";
    process.env.COUCHDB_HOST = "http://localhost:5983";
    process.env.COUCHDB_USER = "env_user";

    await initCmd({ nonInteractive: true });
    fs.accessSync("/tmp/datum/datumrc.yml");
    const configFile = fs.readFileSync("/tmp/datum/datumrc.yml", "utf-8");
    const config = yaml.parseDocument(configFile);
    expect(config.getIn(["connection", "host"])).toEqual(
      process.env.COUCHDB_HOST,
    );
    expect(config.getIn(["connection", "user"])).toEqual(
      process.env.COUCHDB_USER,
    );
  });

  it("uses direct arguments with precendence even over environment variables", async () => {
    process.env.XDG_CONFIG_HOME = "/tmp";
    process.env.COUCHDB_HOST = "http://localhost:5983";
    process.env.COUCHDB_USER = "env_user";

    await initCmd({
      nonInteractive: true,
      host: "http://localhost:5984",
      user: "arg_user",
    });
    fs.accessSync("/tmp/datum/datumrc.yml");
    const configFile = fs.readFileSync("/tmp/datum/datumrc.yml", "utf-8");
    const config = yaml.parseDocument(configFile);
    expect(config.getIn(["connection", "host"])).toEqual(
      "http://localhost:5984",
    );
    expect(config.getIn(["connection", "user"])).toEqual("arg_user");
  });
});
