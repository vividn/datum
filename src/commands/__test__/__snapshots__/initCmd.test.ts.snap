// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initCmd writes datumrc into the xdg_config directory 1`] = `
{
  "connection": {
    "host": "%MEMORY%",
    "password": "password",
    "user": "user",
  },
  "db": "test",
  "project_dir": "/tmp",
}
`;
